# Phase 2 Code Challenge

This code challenge is designed to test your understanding of the Phase 2 material. It covers:

- SQL
- Normal Distribution
- Statistical Tests

_Read the instructions carefully_. You will be asked both to write code and to answer short answer questions.

## Code Tests

We have provided some code tests for you to run to check that your work meets the item specifications. Passing these tests does not necessarily mean that you have gotten the item correct - there are additional hidden tests. However, if any of the tests do not pass, this tells you that your code is incorrect and needs changes to meet the specification. To determine what the issue is, read the comments in the code test cells, the error message you receive, and the item instructions.

## Short Answer Questions

For the short answer questions...

* _Use your own words_. It is OK to refer to outside resources when crafting your response, but _do not copy text from another source_.

* _Communicate clearly_. We are not grading your writing skills, but you can only receive full credit if your teacher is able to fully understand your response.

* _Be concise_. You should be able to answer most short answer questions in a sentence or two. Writing unnecessarily long answers increases the risk of you being unclear or saying something incorrect.

In [4]:
# CodeGrade step0
# Run this cell without changes to import the necessary libraries

import itertools
import numpy as np
import pandas as pd
from numbers import Number
import sqlite3
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

import pickle

---
## Part 1: SQL [Suggested time: 20 minutes]
---
In this part, you will create and execute three SQL queries on the Chinook database. For this challenge **you will need to access the `Album` and `Artist` tables**.

### 1) Connect to the Database.

In [2]:
# CodeGrade step1
# Replace None with appropriate code
# Connect to the Database here ("Chinook_Sqlite.sqlite")

path = "/content/Chinook_Sqlite.sqlite"
conn = sqlite3.connect(path)

In [3]:
assert type(path) == str

In [5]:
# Run this cell without changes to see all the
# tables in the database.

df = pd.read_sql(
    """
    SELECT *
    FROM sqlite_master
    """
, conn
)

df[df['type'] == 'table']

type           name       tbl_name  rootpage  \
0   table          Album          Album         2   
1   table         Artist         Artist         3   
2   table       Customer       Customer         4   
3   table       Employee       Employee         7   
4   table          Genre          Genre         9   
5   table        Invoice        Invoice        10   
6   table    InvoiceLine    InvoiceLine        12   
7   table      MediaType      MediaType        14   
8   table       Playlist       Playlist        15   
9   table  PlaylistTrack  PlaylistTrack        16   
11  table          Track          Track        19   

                                                  sql  
0   CREATE TABLE [Album]\n(\n    [AlbumId] INTEGER...  
1   CREATE TABLE [Artist]\n(\n    [ArtistId] INTEG...  
2   CREATE TABLE [Customer]\n(\n    [CustomerId] I...  
3   CREATE TABLE [Employee]\n(\n    [EmployeeId] I...  
4   CREATE TABLE [Genre]\n(\n    [GenreId] INTEGER...  
5   CREATE TABLE [Invoice]\n(\n    [InvoiceId] INT...  
6   CREATE TABLE [InvoiceLine]\n(\n    [InvoiceLin...  
7   CREATE TABLE [MediaType]\n(\n    [MediaTypeId]...  
8   CREATE TABLE [Playlist]\n(\n    [PlaylistId] I...  
9   CREATE TABLE [PlaylistTrack]\n(\n    [Playlist...  
11  CREATE TABLE [Track]\n(\n    [TrackId] INTEGER...

### 2) Write a query to return the last ten artists alphabetically.

In [19]:
# CodeGrade step2
# Replace None with appropriate code
# Hint: Use the Artist table!

first_query = """
SELECT Name
FROM Artist
ORDER BY Name DESC
LIMIT 10;
"""

pd.read_sql(first_query, conn)

Name
0                     Zeca Pagodinho
1                     Youssou N'Dour
2                           Yo-Yo Ma
3                     Yehudi Menuhin
4                                Xis
5                     Wilhelm Kempff
6                         Whitesnake
7           Vinícius E Qurteto Em Cy
8             Vinícius E Odette Lara
9  Vinícius De Moraes & Baden Powell

In [20]:
# first_query should be a string
assert type(first_query) == str

# first_query should be a SQL query
first_query_df = pd.read_sql(first_query, conn)

### 3) Write a query to return all the albums in the database from Led Zeppelin. Just display the album titles.

In [23]:
second_query = """
SELECT*
FROM Artist  """

pd.read_sql(second_query, conn)

ArtistId                                               Name
0           1                                              AC/DC
1           2                                             Accept
2           3                                          Aerosmith
3           4                                  Alanis Morissette
4           5                                    Alice In Chains
..        ...                                                ...
270       271   Mela Tenenbaum, Pro Musica Prague & Richard Kapp
271       272                             Emerson String Quartet
272       273  C. Monteverdi, Nigel Rogers - Chiaroscuro; Lon...
273       274                                      Nash Ensemble
274       275                              Philip Glass Ensemble

[275 rows x 2 columns]

In [24]:
# CodeGrade step3
# Replace None with appropriate code
# Hint: Use the Artist and Album tables!

second_query = """
SELECT Title
FROM Album
JOIN Artist
  USING(ArtistId)
WHERE Artist.name = "Led Zeppelin";
 """

pd.read_sql(second_query, conn)

Title
0         BBC Sessions [Disc 1] [Live]
1           Physical Graffiti [Disc 1]
2         BBC Sessions [Disc 2] [Live]
3                                 Coda
4                   Houses Of The Holy
5              In Through The Out Door
6                                   IV
7                       Led Zeppelin I
8                      Led Zeppelin II
9                     Led Zeppelin III
10          Physical Graffiti [Disc 2]
11                            Presence
12  The Song Remains The Same (Disc 1)
13  The Song Remains The Same (Disc 2)

In [25]:
# second_query should be a string
assert type(second_query) == str

# second_query should be a SQL query
second_query_df = pd.read_sql(second_query, conn)

### 4) Write a query to return both the artist with the most albums in the database and the respective number of albums they have (1 row).

In [28]:
# CodeGrade step4
# Replace None with appropriate code

third_query = """SELECT Artist.Name , COUNT(Album.AlbumId) AS AlbumCount
FROM Album
JOIN Artist
  USING(ArtistId)
GROUP BY Artist.Name
ORDER BY AlbumCount DESC
LIMIT 1;
"""

pd.read_sql(third_query, conn)

Name  AlbumCount
0  Iron Maiden          21

In [29]:
# third_query should be a string
assert type(third_query) == str

# third_query should be a SQL query
third_query_df = pd.read_sql(third_query, conn)

---
## Part 2: Normal Distribution [Suggested time: 20 minutes]
---
In this part, you will analyze check totals at a TexMex restaurant. We know that the population distribution of check totals for the TexMex restaurant is normally distributed with a mean of $20 and a standard deviation of $3.

### 5) Create a numeric variable `z_score_26` containing the z-score for a $26 check.

In [30]:
# CodeGrade step5
# Replace None with appropriate code
check_amount = 26
mu = 20
sigma = 3
z_score_26 = (check_amount - mu) / sigma
z_score_26

2.0

In [31]:
# This test confirms that you have created a numeric variable named z_score_26

assert isinstance(z_score_26, Number)

### 6) Create a numeric variable `p_under_26` containing the approximate proportion of all checks that are less than or equal to $26.

Hint: Use the answer from the previous question along with the empirical rule, a Python function, or this [z-table](https://www.math.arizona.edu/~rsims/ma464/standardnormaltable.pdf).

In [33]:
# CodeGrade step6
# Replace None with appropriate code
from scipy.stats import norm
p_under_26 =  norm.cdf(z_score_26)
p_under_26

np.float64(0.9772498680518208)

In [34]:
# This test confirms that you have created a numeric variable named p_under_26

assert isinstance(p_under_26, Number)

# These tests confirm that p_under_26 is a value between 0 and 1

assert p_under_26 >= 0
assert p_under_26 <= 1

### 7-A) Create numeric variables `conf_low` and `conf_high` containing the lower and upper bounds (respectively) of a 95% confidence interval for the mean of one waiter's check amounts using the information below.

One week, a waiter gets 100 checks with a mean of $19 and a standard deviation of $3.

In [37]:
# CodeGrade step7
# Replace None with appropriate code

n = 100
mean = 19
std = 3
from scipy.stats import norm
from math import sqrt

z_critical = norm.ppf(0.975)  # For a 95% confidence level

conf_low = mean - z_critical * (std / sqrt(n))
conf_high = mean + z_critical * (std / sqrt(n))


In [38]:
# These tests confirm that you have created numeric variables named conf_low and conf_high

assert isinstance(conf_low, Number)
assert isinstance(conf_high, Number)

# This test confirms that conf_low is below conf_high

assert conf_low < conf_high

# These statements print your answers for reference to help answer the next question

print('The lower bound of the 95% confidence interval is {}'.format(conf_low))
print('The upper bound of the 95% confidence interval is {}'.format(conf_high))

The lower bound of the 95% confidence interval is 18.412010804637983
The upper bound of the 95% confidence interval is 19.587989195362017


### 7-B) Short Answer: Interpret the 95% confidence interval you just calculated in Question 2.3.

# Your answer here
It means we're 95% sure the waiter's true average check amount is between about \$18.41 and \$19.59, based on that week's data. Simple as that.



---
## Part 3: Statistical Testing [Suggested time: 20 minutes]
---
The TexMex restaurant recently introduced queso to its menu.

We have a random sample containing 2000 check totals, all from different customers: 1000 check totals for orders without queso ("no queso") and 1000 check totals for orders with queso ("queso").

### 8-A) Short Answer: State null and alternative hypotheses to use for testing whether customers who order queso spend different amounts of money from customers who do not order queso.

# Your answer here
Null Hypothesis (H₀): Customers who order queso spend the same amount, on average, as those who don't.

Alternative Hypothesis (H₁):Customers who order queso spend a different amount, on average, than those who don't.



### 8-B) Short Answer: What would it mean to make a Type I error for this specific hypothesis test?

Your answer should be _specific to this context,_  not a general statement of what Type I error is.

# Your answer here
It would mean you concluded that queso orders lead to different spending — but in reality, queso doesn’t affect how much customers spend. You falsely saw a difference that isn’t truly there.

### 8-C) Create a numeric variable `p_value` containing the p-value associated with a statistical test of your hypotheses.

In the cell below, we load the sample data for you into the arrays `no_queso` and `queso` for the "no queso" and "queso" order check totals, respectively.

You must identify and implement the correct statistical test for this scenario. You can assume the two samples have equal variances.

Hint: Use `scipy.stats` to calculate the answer - it has already been imported as `stats`. Relevant documentation can be found [here](https://docs.scipy.org/doc/scipy/reference/stats.html#statistical-tests).

In [42]:
# CodeGrade step8
# Load the sample data
no_queso = pickle.load(open('/content/queso.pkl', 'rb'))
queso = pickle.load(open('./queso.pkl', 'rb'))

# Replace None with appropriate code
p_value = stats.ttest_ind(no_queso, queso, equal_var=True).pvalue
p_value


np.float64(1.0)

In [43]:
# This test confirms that you have created a numeric variable named p_value

assert isinstance(p_value, Number)

### 8-D) Short Answer: Can you reject the null hypothesis using a significance level of $\alpha$ = 0.05? Explain why or why not.

# Your answer here
we can’t reject the null hypothesis here. The p-value is 1.0, which is way above the 0.05 threshold. That means there's no evidence of a difference in average spending between queso and no queso customers in this sample.